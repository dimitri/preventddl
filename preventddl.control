# preventddl
default_version = '0.1'
relocatable = 'true'
module_pathname = '$libdir/preventddl'
