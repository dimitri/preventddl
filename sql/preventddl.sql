create extension preventddl;

-- ensure the GUC definition is loaded
load 'preventddl';

show preventddl.enable;

select name, setting, vartype, context
  from pg_settings
 where name = 'preventddl.enable';

set preventddl.enable to true;

-- should fail
create table foo(f1 int, f2 text);

set preventddl.enable to false;

-- should work now
create table foo(f1 int, f2 text);
