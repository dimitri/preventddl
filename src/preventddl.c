/*-------------------------------------------------------------------------
 *
 * preventddl.c
 *	  Register our GUC properly so that it's listed in pg_settings etc.
 *
 * Copyright (c) Citus Data, Inc.
 *-------------------------------------------------------------------------
 */

#include "postgres.h"
#include "utils/guc.h"

PG_MODULE_MAGIC;

#define GUC_STANDARD 0

bool EnablePreventDDL = false;

void _PG_init(void);

void
_PG_init(void)
{
	DefineCustomBoolVariable(
		"preventddl.enable",
		gettext_noop("Enable DDL prevention."),
		gettext_noop("When enabled, DDL commands are prevented and raise "
					 "an exception instead of doing what they are expected to."),
		&EnablePreventDDL,
		false,
		PGC_USERSET,
		GUC_STANDARD,
		NULL, NULL, NULL);
}
