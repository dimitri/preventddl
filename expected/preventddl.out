create extension preventddl;
-- ensure the GUC definition is loaded
load 'preventddl';
show preventddl.enable;
 preventddl.enable 
-------------------
 off
(1 row)

select name, setting, vartype, context
  from pg_settings
 where name = 'preventddl.enable';
       name        | setting | vartype | context 
-------------------+---------+---------+---------
 preventddl.enable | off     | bool    | user
(1 row)

set preventddl.enable to true;
-- should fail
create table foo(f1 int, f2 text);
ERROR:  prevent ddl is enabled, command not allowed
CONTEXT:  PL/pgSQL function preventddl.no_ddl() line 7 at RAISE
set preventddl.enable to false;
-- should work now
create table foo(f1 int, f2 text);
